### PR TITLE
[AMDGPU] - Add address space for strided buffers

### PR DESCRIPTION
**Pull request by OutOfCache**
This is an experimental address space for strided buffers. These buffers can have structs as elements and
a stride > 1.
These pointers allow the indexed access in units of stride, i.e., they point at `buffer[index * stride]`.
Thus, we can use the `idxen` modifier for buffer loads.

We assign address space 9 to 192-bit buffer pointers which contain a 128-bit descriptor, a 32-bit offset and a 32-bit index. Essentially, they are fat buffer pointers with an additional 32-bit index.

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/vectorize-buffer-fat-pointer.ll [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ; RUN: opt -S -mtriple=amdgcn-- -passes=load-store-vectorizer < %s | FileCheck -check-prefix=OPT %s
```
Comment (OutOfCache) [inline llvm/test/CodeGen/AMDGPU/vectorize-buffer-fat-pointer.ll:1]: Should I rename the file? Left it as is for now to see actual changes more easily.

File: llvm/docs/AMDGPUUsage.rst [hunk 2]
```diff
@@ -864,6 +865,17 @@ supported for the ``amdgcn`` target.
   (bits `127:96`). The specific interpretation of these fields varies by the
   target architecture and is detailed in the ISA descriptions.
 
+**Buffer Strided Pointer**
+  The buffer index pointer is an experimental address space. It is supposed to
+  model a 128-bit buffer descriptor and a 32-bit offset, like the **Buffer Fat
+  Pointer**. Additionally, it contains an index into the descriptor, which
+  allows the direct addressing of structured elements.
+
+  The buffer descriptor must be *raw*:
```
Comment (krzysz00) [inline llvm/docs/AMDGPUUsage.rst:874]: I wouldn't describe this as "raw"

Comment (krzysz00) [reply llvm/docs/AMDGPUUsage.rst:874]: And why the requirement for no swizzling?

**Review by krzysz00 — COMMENTED** [review]
How do you intend to rewrite these operations down to the underlying instructions? 

That is, what's your planned equivalent to https://reviews.llvm.org/D158463 ?

**Comment by OutOfCache** [comment]
> How do you intend to rewrite these operations down to the underlying instructions?
> 
> That is, what's your planned equivalent to https://reviews.llvm.org/D158463 ?

Thank you for the link to the code review, I was not aware of your changes before. Up until now, we intended to use the address space in llpc exclusively and lower the pointers to buffer load instructions in `PatchBufferOp`. If we also need to lower these pointers like you did for the fat buffers, maybe we can extend your new pass to also handle these strided buffer pointers?

File: llvm/docs/AMDGPUUsage.rst [hunk 3]
```diff
@@ -864,6 +865,17 @@ supported for the ``amdgcn`` target.
   (bits `127:96`). The specific interpretation of these fields varies by the
   target architecture and is detailed in the ISA descriptions.
 
+**Buffer Strided Pointer**
+  The buffer index pointer is an experimental address space. It is supposed to
+  model a 128-bit buffer descriptor and a 32-bit offset, like the **Buffer Fat
+  Pointer**. Additionally, it contains an index into the descriptor, which
```
Comment (nhaehnle) [inline llvm/docs/AMDGPUUsage.rst:871]: index into the **buffer**

File: llvm/docs/AMDGPUUsage.rst (hunk 2 above)
Comment (nhaehnle) [reply llvm/docs/AMDGPUUsage.rst:874]: It's the same language as for addrspace 7, and I think it makes sense: it's the raw bits as expected by the hardware.

I recall that the buffer intrinsics have a `swz` bit in their flags -- probably due to some HW issue at some point? I don't remember the details.

File: llvm/docs/AMDGPUUsage.rst [hunk 4]
```diff
@@ -864,6 +865,17 @@ supported for the ``amdgcn`` target.
...
+  allows the direct addressing of structured elements.
+
+  The buffer descriptor must be *raw*:
+  the stride is the size of a structured element, the "add tid" flag must be 0, the
+  swizzle eneable bits must be off, and the extent (NumRecords) must be measured in
```
Comment (nhaehnle) [inline llvm/docs/AMDGPUUsage.rst:876]: Typo: enable

Also, drop the part about the extent (NumRecords) -- that's not a requirement of the compiler, it's a requirement of the HW and should be implied by stating that the buffer descriptor is "raw". (And it is also incorrect on IIRC gfx8, which had NUM_RECORDS always in units of bytes)

File: llvm/test/CodeGen/AMDGPU/vectorize-buffer-fat-pointer.ll (hunk 1 above)
Comment (nhaehnle) [reply llvm/test/CodeGen/AMDGPU/vectorize-buffer-fat-pointer.ll:1]: I think it's fine. It's still a kind of buffer fat pointer in a sense.

**Review by nhaehnle — COMMENTED** [review]
Thanks!

Since the plan for now is to have the lowering in LLPC, it's unclear to me how many of the codegen-related changes we actually want in here, though I suppose they don't hurt.

**Comment by krzysz00** [comment]
I'm going to ask the annoying questions:
1. Isn't a strided buffer one where the field that's named something like `stride` (bits 61:48 or 63:48) is non-zero
2. And therefore it uses structured buffers and the `llvm.struct[.ptr].buffer.*` intrinsics?
3. So, with LLVM's gep, how do you tell which field you're GEPing on to?
4. Therefore, is 32-bit index length the right one?

**Comment by nhaehnle** [comment]
>     1. Isn't a strided buffer one where the field that's named something like `stride` (bits 61:48 or 63:48) is non-zero

Yes, and one that is meant to use with buffer instructions using `idxen`.

>     2. And therefore it uses structured buffers and the `llvm.struct[.ptr].buffer.*` intrinsics?

Yes.

>     3. So, with LLVM's gep, how do you tell which field you're GEPing on to?

The intended use is that GEPs affect only the offset part of the fat pointer. The index part is changed using a separate instruction/operation/intrinsic.

>     4. Therefore, is 32-bit index length the right one?

Yes, because that's the index size in the underlying HW.

**Review by nhaehnle — APPROVED** [review]
FWIW, this LGTM. Give it another day or two for other feedback.

File: llvm/docs/AMDGPUUsage.rst [hunk 5]
```diff
@@ -864,6 +865,16 @@ supported for the ``amdgcn`` target.
   (bits `127:96`). The specific interpretation of these fields varies by the
   target architecture and is detailed in the ISA descriptions.
 
+**Buffer Strided Pointer**
+  The buffer index pointer is an experimental address space. It is supposed to
+  model a 128-bit buffer descriptor and a 32-bit offset, like the **Buffer Fat
+  Pointer**. Additionally, it contains an index into the buffer, which
+  allows the direct addressing of structured elements.
```
Comment (krzysz00) [inline llvm/docs/AMDGPUUsage.rst:872]: Document the order of the fields please

**Review by krzysz00 — APPROVED** [review]
Looks good to me, aside from a documentation nit.

**Pull request merged**